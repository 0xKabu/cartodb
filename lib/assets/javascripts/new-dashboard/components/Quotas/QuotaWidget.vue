<template>
  <div class="quota-widget">
    <div class="info-container">
      <h4 class="title is-caption is-txtGrey card-title">{{name}} <span class="is-regular">{{amount}}</span></h4>
      <p class="text is-subheader is-txtGrey" :class="{'is-subheader': !isCompact, 'is-caption': isCompact}">{{roundOneDecimal(usedCapacity)}} / {{roundOneDecimal(availableCapacity)}} {{unit}}</p>
    </div>
    <div class="progressbar">
        <div :class="`progressbar progressbar--${getStatusBar}`"  :style="{width: `${getUsedPercent}%`}">
        </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'QuotaCard',
  props: {
    name: String,
    amount: String,
    availableCapacity: Number,
    usedCapacity: Number,
    unit: String,
    type: String
  },
  computed: {
    getUsedPercent () {
      if (this.availableCapacity === 0) {
        return 100;
      }
      const width = (this.usedCapacity / this.availableCapacity) * 100;
      return width;
    },
    getStatusBar () {
      const usedPercent = this.getUsedPercent;
      if (usedPercent > 90) {
        return 'problem';
      } else if (usedPercent > 60) {
        return 'warning';
      } else {
        return 'good';
      }
    },
    isCompact () {
      return this.type === 'compact';
    }
  },
  methods: {
    roundOneDecimal (number) {
      return Math.round(number * 10) / 10;
    }
  }
};
</script>

<style scoped lang="scss">
@import 'stylesheets/new-dashboard/variables';

.card-title {
  margin-bottom: 8px;
}

.progressbar {
  width: 100%;
  max-width: 100%;
  height: 8px;
  margin-top: 24px;
  border-radius: 4px;
  background-color: $light-grey;

  &.progressbar--good {
    background-color: $good-state;
  }

  &.progressbar--warning {
    background-color: $warning-state;
  }

  .progressbar--problem {
    background-color: $error-state;
  }
}

.quota-widget:not(:only-child) {
  height: auto;
  margin-bottom: 28px;
  padding: 0;
  border: none;
  box-shadow: none;

  &:last-of-type {
    margin-bottom: 0;
  }

  .info-container {
    display: flex;
    justify-content: space-between;
  }

  .card-title {
    margin-bottom: 0;
  }

  .progressbar {
    margin-top: 16px;
  }
}
</style>
