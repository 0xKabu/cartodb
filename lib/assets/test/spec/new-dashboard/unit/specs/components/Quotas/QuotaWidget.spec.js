import { shallowMount } from '@vue/test-utils';
import QuotaWidget from 'new-dashboard/components/Quotas/QuotaWidget';

describe('QuotaWidget.vue', () => {
  it('should render correct contents', () => {
    const quotaWidget = shallowMount(QuotaWidget, {
      propsData: {
        name: 'Widget test',
        amount: 'quota',
        availableCapacity: 1234,
        usedCapacity: 40,
        unit: 'Gb' }
    });

    expect(quotaWidget).toMatchSnapshot();
  });

  it('should put as maximum 100% width in the quota bar', () => {
    const quotaWidget = shallowMount(QuotaWidget, {
      propsData: {
        name: 'Another test',
        amount: 'credits',
        availableCapacity: 1234,
        usedCapacity: 2000
      }
    });

    expect(quotaWidget).toMatchSnapshot();
  });

  it('should render correctly compact widget', () => {
    const quotaWidget = shallowMount(QuotaWidget, {
      propsData: {
        name: 'Another test',
        amount: 'credits',
        availableCapacity: 1234,
        usedCapacity: 12,
        type: 'compact'
      }
    });

    expect(quotaWidget).toMatchSnapshot();
  });

  it('should render warning bar if used ammount is over 60% and under 90%', () => {
    const quotaWidget = shallowMount(QuotaWidget, {
      propsData: {
        name: 'My test',
        amount: 'test',
        availableCapacity: 1000,
        usedCapacity: 610
      }
    });

    expect(quotaWidget).toMatchSnapshot();
  });

  it('should render problem bar if used ammount is over 90%', () => {
    const quotaWidget = shallowMount(QuotaWidget, {
      propsData: {
        name: 'My test',
        amount: 'test',
        availableCapacity: 1000,
        usedCapacity: 932
      }
    });

    expect(quotaWidget).toMatchSnapshot();
  });

  describe('Computed', () => {
    it('should return correct percent', () => {
      const quotaWidget = shallowMount(QuotaWidget, {
        propsData: {
          name: 'My test',
          amount: 'test',
          availableCapacity: 350,
          usedCapacity: 224
        }
      });

      const percent = quotaWidget.vm.getUsedPercent;

      expect(percent).toBe(64);
    });

    it('should not crash and return 100 when the available capacity is 0', () => {
      const quotaWidget = shallowMount(QuotaWidget, {
        propsData: {
          name: 'My test',
          amount: 'test',
          availableCapacity: 0,
          usedCapacity: 0
        }
      });

      const percent = quotaWidget.vm.getUsedPercent;

      expect(percent).toBe(100);
    });

    it('should return correct status of the quota', () => {
      const quotaWidgetGood = shallowMount(QuotaWidget, {
        propsData: {
          name: 'My test',
          amount: 'test',
          availableCapacity: 100,
          usedCapacity: 50
        }
      });
      const quotaWidgetWarning = shallowMount(QuotaWidget, {
        propsData: {
          name: 'My test',
          amount: 'test',
          availableCapacity: 100,
          usedCapacity: 61.4
        }
      });
      const quotaWidgetProblem = shallowMount(QuotaWidget, {
        propsData: {
          name: 'My test',
          amount: 'test',
          availableCapacity: 100,
          usedCapacity: 90.1
        }
      });

      const statusGood = quotaWidgetGood.vm.getStatusBar;
      const statusWarning = quotaWidgetWarning.vm.getStatusBar;
      const statusProblem = quotaWidgetProblem.vm.getStatusBar;

      expect(statusGood).toBe('good');
      expect(statusWarning).toBe('warning');
      expect(statusProblem).toBe('problem');
    });

    it('should return if is compact or not', () => {
      const normalWidget = shallowMount(QuotaWidget, {
        propsData: {
          name: 'My test',
          amount: 'test',
          availableCapacity: 100,
          usedCapacity: 50
        }
      });
      const compactWidget = shallowMount(QuotaWidget, {
        propsData: {
          name: 'My test',
          amount: 'test',
          availableCapacity: 100,
          usedCapacity: 50,
          type: 'compact'
        }
      });

      const normal = normalWidget.vm.isCompact;
      const compact = compactWidget.vm.isCompact;

      expect(normal).toBe(false);
      expect(compact).toBe(true);
    });
  });

  describe('Methods', () => {
    it('should round decimals correctly correct percent', () => {
      const quotaWidget = shallowMount(QuotaWidget, {
        propsData: {
          name: 'My test',
          amount: 'test',
          availableCapacity: 350,
          usedCapacity: 224
        }
      });

      const value1 = quotaWidget.vm.roundOneDecimal(0);
      const value2 = quotaWidget.vm.roundOneDecimal(0.1234);
      const value3 = quotaWidget.vm.roundOneDecimal(120.2);
      const value4 = quotaWidget.vm.roundOneDecimal(129.0);
      const value5 = quotaWidget.vm.roundOneDecimal(45.8);
      const value6 = quotaWidget.vm.roundOneDecimal(45.26);

      expect(String(value1)).toBe('0');
      expect(String(value2)).toBe('0.1');
      expect(String(value3)).toBe('120.2');
      expect(String(value4)).toBe('129');
      expect(String(value5)).toBe('45.8');
      expect(String(value6)).toBe('45.3');
    });
  });
});
