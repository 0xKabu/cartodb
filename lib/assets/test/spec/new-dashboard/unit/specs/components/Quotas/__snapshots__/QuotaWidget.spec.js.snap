// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuotaWidget.vue should render correct contents 1`] = `
<div class="quota-widget">
  <div class="info-container">
    <h4 class="title is-caption is-txtGrey card-title">Widget test <span class="is-regular">quota</span></h4>
    <p class="text is-subheader is-txtGrey is-subheader">40 / 1234 Gb</p>
  </div>
  <div class="progressbar">
    <div class="progressbar progressbar--good" style="width: 3.2414910858995136%;"></div>
  </div>
</div>
`;

exports[`QuotaWidget.vue should render correctly compact widget 1`] = `
<div class="quota-widget">
  <div class="info-container">
    <h4 class="title is-caption is-txtGrey card-title">Another test <span class="is-regular">credits</span></h4>
    <p class="text is-subheader is-txtGrey is-caption">12 / 1234 </p>
  </div>
  <div class="progressbar">
    <div class="progressbar progressbar--good" style="width: 0.9724473257698543%;"></div>
  </div>
</div>
`;

exports[`QuotaWidget.vue should render problem bar if used ammount is over 90% 1`] = `
<div class="quota-widget">
  <div class="info-container">
    <h4 class="title is-caption is-txtGrey card-title">My test <span class="is-regular">test</span></h4>
    <p class="text is-subheader is-txtGrey is-subheader">932 / 1000 </p>
  </div>
  <div class="progressbar">
    <div class="progressbar progressbar--problem" style="width: 93.2%;"></div>
  </div>
</div>
`;

exports[`QuotaWidget.vue should render warning bar if used ammount is over 60% and under 90% 1`] = `
<div class="quota-widget">
  <div class="info-container">
    <h4 class="title is-caption is-txtGrey card-title">My test <span class="is-regular">test</span></h4>
    <p class="text is-subheader is-txtGrey is-subheader">610 / 1000 </p>
  </div>
  <div class="progressbar">
    <div class="progressbar progressbar--warning" style="width: 61%;"></div>
  </div>
</div>
`;
